File: ui/src-tauri/src/commands/deploy.rs
use super::{bridge_call, CommandResponse};
use crate::bridge::BridgeManager;
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::sync::Arc;
use tauri::State;

#[derive(Debug, Serialize, Deserialize)]
pub struct DeployStatus {
    pub environment: String,
    pub host: Option<String>,
    pub services: Vec<ServiceStatus>,
    pub last_deploy: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct ServiceStatus {
    pub name: String,
    pub image: String,
    pub replicas: String,
    pub status: String, // "running" | "stopped" | "partial" | "unknown"
    pub last_updated: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DeployResult {
    pub success: bool,
    pub deployed: u32,
    pub failed: u32,
    pub dry_run: bool,
    pub results: Vec<DeployServiceResult>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct DeployServiceResult {
    pub service: String,
    pub status: String, // "deployed" | "failed" | "would_deploy" | "skipped"
    pub image: Option<String>,
    pub error: Option<String>,
}

/// Get deployment status
#[tauri::command]
pub fn get_deploy_status(
    bridge: State<Arc<BridgeManager>>,
    project_path: String,
    environment: String,
    service: Option<String>,
) -> CommandResponse<Value> {
    match bridge_call(
        &bridge,
        "deploy.status",
        Some(json!({
            "path": project_path,
            "environment": environment,
            "service": service
        })),
    ) {
        Ok(data) => CommandResponse::ok(data),
        Err(e) => CommandResponse::err(e),
    }
}

/// Deploy to environment
#[tauri::command]
pub fn deploy(
    bridge: State<Arc<BridgeManager>>,
    project_path: String,
    environment: String,
    service: Option<String>,
    migrate: bool,
    dry_run: bool,
) -> CommandResponse<Value> {
    match bridge_call(
        &bridge,
        "deploy.deploy",
        Some(json!({
            "path": project_path,
            "environment": environment,
            "service": service,
            "migrate": migrate,
            "dry_run": dry_run
        })),
    ) {
        Ok(data) => CommandResponse::ok(data),
        Err(e) => CommandResponse::err(e),
    }
}

/// Rollback deployment
#[tauri::command]
pub fn rollback_deploy(
    bridge: State<Arc<BridgeManager>>,
    project_path: String,
    environment: String,
    service: Option<String>,
) -> CommandResponse<Value> {
    match bridge_call(
        &bridge,
        "deploy.rollback",
        Some(json!({
            "path": project_path,
            "environment": environment,
            "service": service
        })),
    ) {
        Ok(data) => CommandResponse::ok(data),
        Err(e) => CommandResponse::err(e),
    }
}

/// Get deployment logs
#[tauri::command]
pub fn get_deploy_logs(
    bridge: State<Arc<BridgeManager>>,
    project_path: String,
    environment: String,
    service: String,
    tail: Option<u32>,
    follow: bool,
) -> CommandResponse<Value> {
    match bridge_call(
        &bridge,
        "deploy.logs",
        Some(json!({
            "path": project_path,
            "environment": environment,
            "service": service,
            "tail": tail,
            "follow": follow
        })),
    ) {
        Ok(data) => CommandResponse::ok(data),
        Err(e) => CommandResponse::err(e),
    }
}

/// SSH into deployment environment
#[tauri::command]
pub fn get_ssh_command(
    bridge: State<Arc<BridgeManager>>,
    project_path: String,
    environment: String,
    node: Option<String>,
) -> CommandResponse<Value> {
    match bridge_call(
        &bridge,
        "deploy.ssh_command",
        Some(json!({
            "path": project_path,
            "environment": environment,
            "node": node
        })),
    ) {
        Ok(data) => CommandResponse::ok(data),
        Err(e) => CommandResponse::err(e),
    }
}

File: ui/src/api/deploy.ts
/**
 * Deployment API
 */

import { invoke } from '@tauri-apps/api/core';
import type { CommandResponse } from '../types';

export interface ServiceStatus {
  name: string;
  image: string;
  replicas: string;
  status: string; // "running" | "stopped" | "partial" | "unknown"
  last_updated: string | null;
}

export interface DeployStatus {
  environment: string;
  host: string | null;
  services: ServiceStatus[];
  last_deploy: string | null;
  error?: string;
}

export interface DeployResult {
  success: boolean;
  deployed: number;
  failed: number;
  dry_run: boolean;
  requires_approval?: boolean;
  results: Array<{
    service: string;
    status: string;
    image: string | null;
    error: string | null;
  }>;
}

export interface RollbackDeployResult {
  success: boolean;
  rolled_back: number;
  failed: number;
  results: Array<{
    service: string;
    status: string;
    error: string | null;
  }>;
}

export interface DeployLogs {
  service: string;
  environment: string;
  logs: string;
}

/**
 * Get deployment status
 */
export async function getDeployStatus(
  projectPath: string,
  environment: string,
  service?: string
): Promise<DeployStatus> {
  const response = await invoke<CommandResponse<DeployStatus>>('get_deploy_status', {
    projectPath,
    environment,
    service,
  });
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to get deploy status');
  }
  return response.data;
}

/**
 * Deploy to environment
 */
export async function deploy(
  projectPath: string,
  environment: string,
  service?: string,
  migrate: boolean = false,
  dryRun: boolean = false
): Promise<DeployResult> {
  const response = await invoke<CommandResponse<DeployResult>>('deploy', {
    projectPath,
    environment,
    service,
    migrate,
    dryRun,
  });
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to deploy');
  }
  return response.data;
}

/**
 * Rollback deployment
 */
export async function rollbackDeploy(
  projectPath: string,
  environment: string,
  service?: string
): Promise<RollbackDeployResult> {
  const response = await invoke<CommandResponse<RollbackDeployResult>>('rollback_deploy', {
    projectPath,
    environment,
    service,
  });
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to rollback deployment');
  }
  return response.data;
}

/**
 * Get deployment logs
 */
export async function getDeployLogs(
  projectPath: string,
  environment: string,
  service: string,
  tail?: number,
  follow: boolean = false
): Promise<DeployLogs> {
  const response = await invoke<CommandResponse<DeployLogs>>('get_deploy_logs', {
    projectPath,
    environment,
    service,
    tail,
    follow,
  });
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to get deploy logs');
  }
  return response.data;
}

/**
 * Get SSH command for environment
 */
export async function getSshCommand(
  projectPath: string,
  environment: string,
  node?: string
): Promise<{ command: string; user: string; host: string }> {
  const response = await invoke<
    CommandResponse<{ command: string; user: string; host: string }>
  >('get_ssh_command', {
    projectPath,
    environment,
    node,
  });
  if (!response.success || !response.data) {
    throw new Error(response.error || 'Failed to get SSH command');
  }
  return response.data;
}
